arch/common/shared_irq.c

```diff
@@ -22,3 +23,76 @@ void shared_isr(const void *data)
 		}
 	}
 }
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+
+static struct k_spinlock lock;
+
+void z_isr_install(unsigned int irq, void (*routine)(const void *),
+		   const void *param)
+{
+	struct shared_irq_data *irq_data;
+	struct _isr_table_entry *entry;
+	struct shared_irq_client *client;
+	unsigned int table_idx;
+	int i;
+	k_spinlock_key_t key;
+
+	table_idx = get_sw_isr_table_idx(irq);
+
+	/* check for out of bounds table index */
+	if (table_idx >= CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	irq_data = &_shared_sw_isr_table[table_idx];
+	entry = &_sw_isr_table[table_idx];
+
+	/* number of clients shouldn't exceed the client limit */
+	__ASSERT(irq_data->client_num < CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS,
+		 "reached client limit");
+
+	key = k_spin_lock(&lock);
```

Comment:
Not worth a -1 because it's an assert, but note that putting this before the lock introduces a race: it's possible for two (or more, technically) contexts to come through, both check that client_num has space for one more item, and then **both** grab the lock and add **two** items to the array, causing an overflow.

Reply:
@LaurentiuM1234 I think that @andyross's above comment worth to be added in the code as note/comment.

Reply:
Thanks for pointing this out. This was a slip-up on my behalf. Fixed now.